### PR TITLE
Fix incorrect `this` binding in `babel-core` + `babel-traverse`

### PR DESCRIPTION
**Pull request by marvinhagemeister**
This PR fixes incorrect binding of `this` for babel plugins. The `this` object is used to share state across the `pre`, `post` and `visitor`-handlers as can be seen in the following [example in the babel handbook](https://github.com/thejameskyle/babel-handbook/blob/master/translations/en/plugin-handbook.md#-pre-and-post-in-plugins):

```js
export default function({ types: t }) {
  return {
    pre(state) {
      this.cache = new Map();
    },
    visitor: {
      StringLiteral(path) {
        this.cache.set(path.node.value, 1);
      }
    },
    post(state) {
      console.log(this.cache);
    }
  };
}
```

This is done by adding a plugin object interface (a babel plugin is a function which must return this object) and by making `Visitor` generic.

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/thejameskyle/babel-handbook/blob/master/translations/en/plugin-handbook.md#-pre-and-post-in-plugins
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/babel-core/index.d.ts [hunk 1]
```diff
@@ -6,13 +6,26 @@
...
 export { traverse, Visitor };
 
+// A babel plugin is a simple function which must return an object matching
+// the following interface. Babel will throw if it finds unknown properties.
+// The list of allowed plugin keys is here:
+// https://github.com/babel/babel/blob/4e50b2d9d9c376cee7a2cbf56553fe5b982ea53c/packages/babel-core/src/config/option-manager.js#L71
+export interface PluginObj<S = {}> {
+    name?: string;
+    manipulateOptions?(opts: any, parserOpts: any): void;
+    pre?(this: S, state: any): void;
```
Comment (Kovensky) [inline types/babel-core/index.d.ts]: The value of `state` is actually the same as the value of `this`. You can check by doing a `===` comparison -- they are the same value.

This is true of all the plugin keys that receive a `state` as their last parameter, including `Visitor`; which seems to be using an incorrect template argument.

Comment (Kovensky) [inline types/babel-core/index.d.ts]: It also seems that the `state` type is shared with all visitors being visited during a single `traverse` call. I'm not sure it is possible to correctly type it in typescript :/

Comment (marvinhagemeister) [inline types/babel-core/index.d.ts]: @Kovensky Are you sure about that? For me that isn't the case with latest babel. Only inside the visitor functions `state` is equal to `this`.

```js
const babel = require("babel-core");

function plugin() {
  return {
    pre(state) {
      state.foo = "foo";
      console.log("pre:", state.foo, ", equal:", this === state);
    },
    visitor: {
      VariableDeclaration(path, state) {
        console.log("visitor:", state.foo, ", equal:", this === state);
        state.foo = "bar";
      },
    },
    post(state) {
      console.log("post:", state.foo, ", equal:", this === state);
    },
  };
}

babel.transform(`const foo = "Hello World";`, { plugins: [plugin] });

// logs:
pre: foo , equal: false
visitor: undefined , equal: true
post: foo , equal: false
```

Notice how  `state` is never set to `bar`. Now compare this to setting state on `this`:

```js
// const babel = require("babel-core");

// function plugin() {
//   return {
//     pre(state) {
//       state.foo = "foo";
//       console.log("pre:", state.foo, ", equal:", this === state);
//     },
//     visitor: {
//       VariableDeclaration(path, state) {
//         console.log("visitor:", state.foo, ", equal:", this === state);
//         state.foo = "bar";
//       },
//     },
//     post(state) {
//       console.log("post:", state.foo, ", equal:", this === state);
//     },
//   };
// }

// babel.transform(`const foo = "Hello World";`, { plugins: [plugin] });

const babel = require("babel-core");

function plugin() {
  return {
    pre(state) {
      this.foo = "foo";
      console.log("pre:", this.foo, ", equal:", this === state);
    },
    visitor: {
      VariableDeclaration(path, state) {
        console.log("visitor:", this.foo, ", equal:", this === state);
        this.foo = "bar";
      },
    },
    post(state) {
      console.log("post:", this.foo, ", equal:", this === state);
    },
  };
}

babel.transform(`const foo = "Hello World";`, { plugins: [plugin] });

// Logs:
pre: foo , equal: false
visitor: foo , equal: true
post: bar , equal: false
```

My PR fixes exactly that and makes `this` type safe by using the same interface in `pre+post` and `visitor` functions.

**Comment by mhegazy** [comment]
@Kovensky can i get a signoff from you on this change?

**Comment by DanielRosenwasser** [comment]
🔔

File: types/babel-core/index.d.ts (hunk 1 above)
Comment (Kovensky) [inline types/babel-core/index.d.ts]: I wonder what _is_ `this` equivalent to then 🤔 

It's hard to follow in the babel codebase since the `state` is handled inside a closure created by a helper function for creating such closures 🤷‍♀️

**Review by Jessidhia — APPROVED** [review]
Sorry, got lost in all the notifications. Should be ok.
